test/e2e/xfn_test.go

```diff
@@ -0,0 +1,344 @@
+}
+
+// ManagedResourcedProcessedByFunction asserts that MRs contains the requested label
+func ManagedResourcedProcessedByFunction() features.Func {
+
+	return func(ctx context.Context, t *testing.T, config *envconf.Config) context.Context {
+		labelName := "labelizer.xfn.crossplane.io/processed"
+		rg := resourceGetter(ctx, t, config)
+		claim := rg("fn-labelizer", "default", "nop.example.org/v1alpha1", "NopResource")
+		r := resourceValue(t, claim, "spec", "resourceRef")
+
+		xr := rg(r["name"], "default", r["apiVersion"], r["kind"])
+		mrefs := resourceSliceValue(t, xr, "spec", "resourceRefs")
+		for _, mref := range mrefs {
+			err := wait.For(func() (done bool, err error) {
+				mr := rg(mref["name"], "default", mref["apiVersion"], mref["kind"])
+				l, found := mr.GetLabels()[labelName]
+				if !found {
+					return false, nil
+				}
+				if l != "true" {
+					return false, nil
+				}
+				return true, nil
+			}, wait.WithTimeout(5*time.Minute))
+			if err != nil {
+				t.Fatalf("Expected label %v value to be true", labelName)
+			}
+
+		}
+		return ctx
+	}
+}
+
+// CreateTLSCertificateAsSecret for given dns name and store the secret in the given namespace
+func CreateTLSCertificateAsSecret(dnsName string, ns string) features.Func {
+	return func(ctx context.Context, t *testing.T, config *envconf.Config) context.Context {
+		caPem, keyPem, err := createCert(dnsName)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "reg-cert",
+				Namespace: ns,
+			},
+			Type: corev1.SecretTypeTLS,
+			StringData: map[string]string{
+				"tls.crt": caPem,
+				"tls.key": keyPem,
+			},
+		}
+		client := config.Client().Resources()
+		if err := client.Create(ctx, secret); err != nil {
+			t.Fatalf("Cannot create secret %s: %v", secret.Name, err)
+		}
+		configMap := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "reg-ca",
+				Namespace: namespace,
+			},
+			Data: map[string]string{
+				"domain.crt": caPem,
+			},
+		}
+		if err := client.Create(ctx, configMap); err != nil {
+			t.Fatalf("Cannot create config %s: %v", configMap.Name, err)
+		}
+		return ctx
+	}
+}
+
+// CopyImagesToRegistry copies fn images to private registry
+func CopyImagesToRegistry() features.Func {
+	return func(ctx context.Context, t *testing.T, config *envconf.Config) context.Context {
+		nodes := &corev1.NodeList{}
+		if err := config.Client().Resources().List(ctx, nodes); err != nil {
+			t.Fatal("cannot list nodes", err)
+		}
+		if len(nodes.Items) == 0 {
+			t.Fatalf("no nodes in the cluster")
+		}
+		var addr string
+		for _, a := range nodes.Items[0].Status.Addresses {
+			if a.Type == corev1.NodeInternalIP {
+				addr = a.Address
+				break
+			}
+		}
+		if addr == "" {
+			t.Fatalf("no nodes with private address")
+		}
+
+		srcRef, err := name.ParseReference("crossplane-e2e/fn-labelizer:latest")
+		if err != nil {
+			t.Fatal(err)
+		}
+		src, err := daemon.Image(srcRef)
```

Comment:
Can't you instead use crane.Copy?

Reply:
Ah, I guess not because the image is only built locally. I was suggesting that not to have to rely on the docker daemon directly

Reply:
> Ah, I guess not because the image is only built locally. I was suggesting that not to have to rely on the docker daemon directly

Correct, `crane.Copy()` can copy only between two remote repos.